### ISSUE TITLE
Thread synchronization is blocket until window is moved or resized

### ISSUE THREAD

**Issue by michalgw** [graemeg/fpGUI#14]
Thread synchronization is blocket until window is moved or resized. This only happens under windows-gdi. On linux-x11 it works fine.

Run test project and try to move or resize the window.
Test project:
http://sourceforge.net/p/fpgui/bugs/15/attachment/testsync.zip

**Comment by fredvs** [comment]
Hello.

I have try all the examples of michalgw with fpc 3.0.0. and fpGUI 1.4 maint with wine and Win10.
It works perfect here.

I try changing:
`Synchronize(@DoSync);`
with
`Queue(@DoSync);`

It works too.

Fre;D

**Comment by rchastain** [comment]
Tested on Linux with latest fpGUI source (_develop_ branch). It works perfectly. I believe that this issue could be closed.

**Comment by michalgw** [comment]
As I wrote, the problem occurs on Windows. It worked fine on Linux.

**Comment by rchastain** [comment]
> As I wrote, the problem occurs on Windows. It worked fine on Linux.

Ah, sorry. I will make a test on Windows ASAP.

**Comment by rchastain** [comment]
Compiled and tested the test program under Windows. Both versions (32-bit and 64-bit) work as expected for me.

Latest fpGUI, _develop_ branch, FPC 3.2.0.

[threadsynctest-binaries-windows.zip](https://github.com/graemeg/fpGUI/files/10663774/threadsynctest-binaries-windows.zip)

**Comment by michalgw** [comment]
I just tested and the problem still persists (fpgui-develop and fpc-main). To reproduce the problem, run program and grab title bar of form - thread synchronization is suspended.

**Comment by fredvs** [comment]
Hello.

Could you give, please, a code-example to test the synchro (here I dont have problems with grabbing the title  bar.) ?
Thanks.

**Comment by michalgw** [comment]
Try examples I've attached earlier:
http://sourceforge.net/p/fpgui/bugs/15/attachment/testsync.zip
http://pastebin.com/mfsx2H9Z

**Comment by rchastain** [comment]
> I just tested and the problem still persists (fpgui-develop and fpc-main). To reproduce the problem, run program and grab title bar of form - thread synchronization is suspended.

Indeed, I didn't try to grab the title bar. I will try tomorrow.

**Comment by fredvs** [comment]
Hello michalgw.

Tested  your demo compiled  with fpc 3.2.2.  64 bit and fpGUI last develop branch on Linux 64 bit.
Tested also  your demo compiled  with fpc 3.2.2.  32 bit and fpGUI last develop branch on Windows 11/last update .
The Linux demo works ok when grabbing the window too.
The Windows demo works ok  when grabbing on Linux via Wine and also on Windows 11/last update.

In attachment, the binary of your demo ( for Linux and Windows).  
[testsync_bin.zip](https://github.com/graemeg/fpGUI/files/10665464/testsync_bin.zip)

Do you have also problems with that binary?

**Comment by michalgw** [comment]
Yes, the problem occurs with that binary (tested on Windows 10). When I grab title bar and move window - the form stops blinking and counter is not incremented until mouse button is released.

**Comment by rchastain** [comment]
> Yes, the problem occurs with that binary (tested on Windows 10). When I grab title bar and move window - the form stops blinking and counter is not incremented until mouse button is released.

Don't feel offended, but shall we really consider that as an issue?

**Comment by fredvs** [comment]
Ha, ok, I get it now.

Indeed, on Windows 11, the form stop blinking while the mouse button is pressed.
This does not occur with Linux-wine ( but Linux-wine is the best Windows version ever ;-) )

**Comment by michalgw** [comment]
> Don't feel offended, but shall we really consider that as an issue?

If you are using threads then definitely yes.


> Ha, ok, I get it now.

I probably should have described the problem better at the beginning.

**Comment by fredvs** [comment]
Interesting is that using _queue()_ instead of _synchronize()_ (https://github.com/graemeg/fpGUI/issues/14#issuecomment-272724697) does not loose the sequence while moving the window (like _synchronize()_ does).

You may try with the binary compiled with _queue()_:
[testsync_queue.zip](https://github.com/graemeg/fpGUI/files/10668462/testsync_queue.zip)

**Comment by michalgw** [comment]
But if you run the second example (http://pastebin.com/mfsx2H9Z) with changed synchronize to queue the window remains unresponsive to mouse and keyboard because windows messages are not processed.

**Comment by fredvs** [comment]
Yes, it is what I noted.

With synchronize, for example if you begin to move the window with number "10" a certain time,  the window is not refreshed and when you release the button, the next number wil be "11".

But with queue, if you begin with number "10" and move the window a certain time, it is not refreshed too but when you release the button, the next number will not be "11" but the number inc during the moving.

**Comment by fredvs** [comment]
> But if you run the second example (http://pastebin.com/mfsx2H9Z) 

Sorry but for all your link to pastebin I get error connection  and cannot test it.

**Comment by fredvs** [comment]
Just tested with a MSEgui app with thread on Windows 11 and there all ok when moving window.

Now where to look, both use Windows gdi32 interface...?
Any idea is welcome.

**Comment by fredvs** [comment]
Re-hello.

> But if you run the second example (http://pastebin.com/mfsx2H9Z)

You may use this issue and paste the zipped file, it is more easy for everybody.

**Comment by michalgw** [comment]
I suspect the main application loop and/or gdi window procedure.

I attach an example:
[onmove.zip](https://github.com/graemeg/fpGUI/files/10669707/onmove.zip)

**Comment by fredvs** [comment]
Hum, without minimum sleep(1) I cannot close your example.
Anyway, yes there is  a problem when moving window and thread-synchronization.

Did you try to synchronize something inside the form, like a paintbox? (sorry to ask it but I dont have easy access to Windows).

Is synchro bad too?

The best is to try to isolate the problem, when it begins (on mousedown, mouseclick, when moving the form, etc...? )

Then check on the TForm.method with problem what it does and maybe compare with msegui.

**Comment by michalgw** [comment]
Indeed, this example with queue is not the best, because the return from CheckSynchronize may never happen - more queue items may arrive while the queue is being processed. 

Using synchronize it works like the first project - it pauses synchronization on moving and resizing.

Does fpGUI have paintbox? I can not find...

You can try to add DebugLn to beginning of fpgWindowProc and dump the windows message to see what's going on while moving/resizing.

**Comment by graemeg** [comment]
I can have a look tomorrow evening, after work. The fpgWindowProc should already have debug code for OS event debugging - you just need to compile the framework with the correct Defines.

Clicking on the title bar will not (should not) trigger a OnMouseDown if I recall, because that's outside the bounds of the application window. When moving the window, that should send WindowMove OS events to the application event handler.

Let me see what I can find. Thank you all for your investigation so far - it is very handy, and should help with debugging.

**Comment by fredvs** [comment]
> Does fpGUI have paintbox? I can not find...

You may try painting on the canvas of a TOtherwidget or TFPGPanel.

**Comment by fredvs** [comment]
> Does fpGUI have paintbox? I can not find...

Or simply, with a TfpgPanel, change the color and caption via the thread-synchronize.

**Comment by michalgw** [comment]
I just tried with TfpgPanel but effect is the same.

**Comment by graemeg** [comment]
I've rewritten the event loop a few weeks ago and threading now wakes the main event loop automatically, to make sure it processes event (not just ones from the windowing system). Tested extensively and it's working correctly now. Thank you for reporting this.
